### PR TITLE
(#147) DIModule

### PR DESCRIPTION
**Pull request by dannypsnl**
ref: llir/grammar#7

### CODE REVIEW

File: asm/specialized_metadata.go [hunk 1]
```diff
@@ -1110,8 +1110,16 @@ func (gen *generator) irDIModule(new metadata.SpecializedNode, old *ast.DIModule
...
+		case *ast.APINotesField:
+			md.APINotes = stringLit(oldField.APINotes())
+		case *ast.FileField:
+			file, err := gen.irMDField(oldField.File())
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			md.File = file
+		case *ast.LineField:
+			*md.Line = intLit(oldField.Line())
```
Comment (mewmew) [inline asm/specialized_metadata.go:1122]: Would this not result in a nil pointer reference? Since `Line` has not yet been allocated.

**Review by mewmew — COMMENTED** [review]
LGTM, with one minor comment. `md.Line` may have to be allocated before assigned?

File: asm/specialized_metadata.go (hunk 1 above)
Comment (dannypsnl) [reply asm/specialized_metadata.go:1122]: I'm so surprised deference left value in Go didn't work just like C...

**Comment by dannypsnl** [comment]
Notice that tests would fail since we haven't update `testdata`, @mewmew would you like to update that? I thought we could also find out some missing cases just like the last time.

**Comment by mewmew** [comment]
> Notice some tests would fail since we haven't update testdata, @mewmew would you like to update that? I thought we could also find out some missing cases just like the last time.

Done. Test cases have now been updated to LLVM 11.0. Created issue #158 to track the progress.

Cheers,
Robin

File: ir/metadata/specialized_metadata.go [hunk 2]
```diff
@@ -1429,7 +1429,9 @@ type DIModule struct {
 	Name         string // required.
 	ConfigMacros string // optional; empty if not present.
 	IncludePath  string // optional; empty if not present.
-	Sysroot      string // optional; empty if not present.
+	APINotes     string // optional; empty if not present.
+	File         Field  // optional; empty if not present.
+	Line         *int64 // optional; empty if not present.
```
Comment (mewmew) [inline ir/metadata/specialized_metadata.go:1434]: As Go is lacking optional values, using pointers is a common solution.

For the specialized metadata nodes, I've tried to rely on zero values for omitted values, where those zero values give no information regarding the source. For instance, line number 0 does not exist, as source line numbers are 1-indexed. For this reason, we can use a regular zero value of `int64` for the Line field.

This is how it was done for `DICommonBlock`:

From https://github.com/llir/llvm/blob/fbfa5b12032936ac00ed05780808a6f917694f78/ir/metadata/specialized_metadata.go#L102
```go
type DICommonBlock struct {
	...
	Line        int64   // optional; zero value if not present.
}
```

The balance here is between correctness and ease of use. Having an API with pointer values for every field makes it more difficult to use.

As both approaches have merit, we should decide on one approach and use it throughout for consistency. So either we change `Line` to be `int64` to be consistent with `DICommonBlock` (and others). Or we change all other places to use pointers as well.

Definitely open for input.

Cheers,
Robin

**Review by mewmew — APPROVED** [review]

**Pull request merged**